DM-45873: Initial query_all_datasets implementation

dhirving opened the pull request:
## Checklist

Extracted the query logic from the `query-datasets` CLI and used it to provide an initial implementation of query_all_datasets.  

This gets the interface and unit tests in place for a future PR that will run the whole query on the server side for RemoteButler.

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

dhirving committed:
Extract dataset query logic from CLI

Pull out the "query all datasets" logic from the query-datasets CLI command to a separate function.  In an upcoming commit this will be used to implement `Butler.query_all_datasets`.

dhirving committed:
Pull out collection filtering logic to a helper function

dhirving committed:
Add initial implementation of query_all_datasets

Add a method for querying multiple dataset types simultaneously, currently hidden as `Butler._query_all_datasets`.

This implementation uses the existing logic from the query-datasets CLI for doing the search.

dhirving committed:
Fix missing default for query-datasets type

The dataset type is supposed to default to '*' if no dataset types are provided by the user.

dhirving committed:
Fix repeated limit warnings

Fix an issue where the limit was being issued repeatedly because the caller was modifying the results array before query_all_datasets checked its length.

timj commented on python/lsst/daf/butler/script/queryDatasets.py:
```diff
@@ -258,83 +260,50 @@ def getDatasets(self) -> Iterator[list[DatasetRef]]:
+                collections=query_collections,
                 find_first=self._find_first,
+                name=self._dataset_type_glob,
                 with_dimension_records=True,
```
Following on from a Slack comment, it looks like there are other scripts using the `QueryDatasets` class that do need dimensions so if you want to be able to turn this off for `query-datasets` but turn it on for `transfer-datasets` then you will need to add a parameter higher up.

dhirving replied:
Thanks -- I failed to notice that `QueryDatasets` was shared by multiple scripts.  I had removed order-by and the dimension records in the follow-up PR to this one but I'll have to put the dimension records back.

timj commented on tests/test_cliCmdQueryDatasets.py:
```diff
@@ -357,6 +357,13 @@ def test_limit_order(self):
+        # Same as previous test, but with positive limit so no warning is
+        # issued.
+        tables = self._queryDatasets(
+            repo=testRepo.butler, limit=1, order_by=("visit"), collections="*", glob="*"
```
Maybe test that no warning was issued (by issuing your own warning and then making sure you only got the one warning)?

timj reviewed (approved):
Thanks. I've tested it and it does seem to work as expected. It looks like the restriction of order-by only working for a single dataset type is not part of this PR yet.

dhirving committed:
Add test for no warnings issued with positive limit

dhirving committed:
Make order-by error more informative

Add a description of which columns are invalid and which are allowed when an order by expression is not legal.
